meps/mep-0004.md

```diff
@@ -0,0 +1,273 @@
+---
+label: MEP0004
+title: Per-Output AST Representation for Code Cell Outputs
+short_title: Per-Output AST
+description: Extend the MyST AST to represent individual code cell outputs as separate AST nodes, enabling output content to participate in MyST builds (e.g., generate and consume reference labels).
+date: 2025/12/04
+authors:
+  - agoose77
+  - rowanc1
+tags:
+  - Draft
+data:
+  discussion: https://github.com/jupyter-book/mystmd/issues/1026
+---
+
+## Summary
+
+We propose extending the MyST AST with a new node type to represent individual code-cell outputs as separate AST nodes. In the existing schema, the `Output` node type represents a collection of outputs, and has a one-to-one correspondence with each code-cell. This structure does not naturally admit distinguishing between AST trees for each output, precluding the ability to consider code-cell outputs during subsequent transformations of the document. With this proposal, a new `Outputs` node replaces the `Output` node as a direct child of code-cells with a unist `Parent` container. This new node may contain several `Output` child nodes, each possessing an `IOutput` bundle returned by code-execution. The new AST structure enables each `Output` to carry its own AST subtree. Subsequent enhancements may build upon this proposal to process code-cell outputs and build derived ASTs, e.g. parsing Markdown or LaTeX outputs into MyST AST that may reference and be referenced by other content.
+
+## Context
+
+### Current Limitations
+
+Code-cells from Jupyter Notebooks can produce rich output like Markdown, LaTeX, and tables. Right now, each output is effectively treated as a <wiki:Black_Box>, which is only interpreted at export time (to PDF, web, etc.) such that static exports and web build are required to interpret the MIME bundle outputs. As a result, the output content does not participate in a MyST build i.e. the Markdown or LaTeX output cannot generate or consume referencing labels.
+
+This limitation prevents programmatic generation of content that integrates with the rest of the document. For example, if a code cell generates Markdown output containing a figure with a label, that label cannot be referenced from other parts of the document, and the figure cannot appear in cross-reference resolution.
+
+### Use Cases
+
+This enhancement enables several important use cases:
```

Comment:
```suggestion
This enhancement lays the foundation to enable several important use cases (it does not solve for those use-cases now, but is a required step):
```

Reply:
I think this is true - we still need to do the follow up work for the other use-cases, right? This just makes that clear